src/protocol/mod.rs

```diff
@@ -231,7 +231,9 @@ impl WebSocketContext {
         Stream: Read + Write,
     {
         // Do not write to already closed connections.
-        self.state.check_active()?;
+        if !self.state.is_active() {
+            return Err(Error::AlreadyClosed);
```

Comment:
Looking at it again, I feel the correct error type to return here might be `Protocol`? What do you think?

Reply:
> Tbh as a user of the library, this isn't clear at all. None of the methods on WebSocket have a comprehensive documentation about what errors might be returned, what they mean and what you should do about them. Even more so when using through tokio-tungstenite which adds another layer of indirection.

Yes, the docs for `AlreadyClosed` are a bit vague, I also don't like this error type to be honest. However things like `ConnectionClosed` had a decent documentation AFAICT.

> It's possible to introduce more specific error types, but it might just make it more confusing.

Exactly, that's the problem. We might have introduced different error types for `write_message()` and `read_message()`, but that would only make things harder, especially when you use the tokio counterpart of this library (current stable tokio version) as that would require always matching the error types / converting them to something like `Box<dyn Error>` when you combine read and write counterparts in `select()`.

> "trying to read after receiving Error::ConnectionClosed": do you receive this error in the reader part or when trying to write? That's also a bit confusing. From doing a search in the source code, it turns out that:

Yes, that might be confusing at the first sight, but normally you only get this message while reading and getting a `ConnectionClosed` allows you to write a graceful shutdown and drop the connection. In terms of `tokio-tungstenite` -- it's used to identify when to return `None` from the `Stream` which normally gives the expected behavior (especially if you use it the way it was shown in examples).

> In 2 it is indeed annoying that the "information" about the connection being closed is in the reader part, which might be a separate async task. It means you somehow have to tell your writer task to stop, requiring some sort of synchronization.
> ...
> Just as a sidenote, another way to detect events might be out of band. I tend to find it a nicer API. Let the object be observable and send an event on (fatal) errors and connection close, so every component in a software can listen for those events if they care and act on them. It might not be a good fit for tungstenite because it is quite low level, but it might be nice for tokio-tungstenite. I wrote the pharos crate for this. It's light on features for now but I will add things like filtering events with a predicate if you only care about some of them.



I'm not sure if I fully understood your point. The current way the `tungstenite-rs` is written is designed was flexible enough and gave us sound invariants to implement the client and server based on `tungstenite-rs` and `tokio-tungstenite`, it might be that for the new tokio version some refactoring would be required. I can't comment on that much as I did not dig into all these details of new tokio (I know some basics based on "This Week in Rust"). Once the new tokio gets stable, so that we can use it in production, I'll migrate `tokio-tungstenite` (unless it is migrated by other guys by that moment) and we'll see if it gives reasonable API for users and for library authors (as far as I understood, you're developing some framework based on `tokio-tungstenite`).

But in any case we would avoid the design issues of `ws-rs` and some event-based approaches as we want to have strong invariants in code and the possibility to write "FRP-like" code.

Reply:
> Looking at it again, I feel the correct error type to return here might be Protocol? Given that ConnectionClosed and AlreadyClosed don't return the message to the user either.

Yes, but the `ConnectionClosed` is _normally_ what you get when you read the message. I think if we just return `AlreadyClosed` the way you wrote it, then `AlreadyClosed` does not make much sense anymore (previously it was triggered only if you tried to do something with the socket which is "already closed", i.e. after receiving `ConnectionReset`).

So I think an error which we return in case someone tries to write a message to a socket after initiating connection closure should be different (I'm referring to the `write_message()` part, the rest seems good to me).

Reply:
> but normally you only get this message while reading
> Yes, but the ConnectionClosed is normally what you get when you read the message.

I feel there is some confusion here about when `ConnectionClosed` get's returned. From checking the code, this can be returned from both `send_one_frame` and `write_pending`, so definitely you get this on sending.

That's why I supposed it was considered a means of telling a sender task that the connection is closed and it should stop trying to send on it. 

So to try and resolve this, this is one option:
```rust
// When terminated, return AlreadyClosed.
self.state.check_active()?;

// Do not write after sending a close frame.
if !self.state.is_active() {
    return Err(Error::Protocol("Sending after closing is not allowed".into()));
}
```

Some considerations: 
- Until now, the only use for `Error::Protocol` was when reading and the remote made a protocol violation, so this introduces something new all together in the API. 
- In theory, here we might not have sent a CloseFrame just yet if the remote initiated the close, but this will call `write_pending` which will answer the close frame, and by the time this message would be sent out, it wouldn't be valid to send anymore, so it's probably best that the API hides that fact and pretends that we respond immediately to keep things consistent. 
  That does mean that client code now really has to check for this in sender tasks, because it can happen right after receiving a close frame, but before receiving `ConnectionClosed`. 

As for sending `ConnectionClosed` here, that would break the guarantee that it's safe to drop the underlying transport once receiving that, because here the handshake is only guaranteed to be half finished (we might be waiting for acknowledgement).

Reply:
> I feel there is some confusion here about when ConnectionClosed get's returned. From checking the code, this can be returned from both send_one_frame and write_pending, so definitely you get this on sending.

Well, for simplicity let's omit `send_one_frame()` as it's a private function and the user does not see it. Your concern is that this error might be returned from both `write_pending()` and `read_message()`, right? While this is true, in most cases (if not all) `write_pending()` returns `ConnectionClosed` when called inside `read_message()` (as this function calls `write_message()` internally).

Especially if we take into consideration the changes in this PR and the fact that sending message after sending `Close` is an error... I think it will be barely no cases when you can get `ConnectionClosed` from calling `write_pending()` explicitly.

> So to try and resolve this, this is one option:

Yes, that looks good for me.

> Some considerations:
> Until now, the only use for Error::Protocol was when reading and the remote made a protocol violation, so this introduces something new all together in the API.

Yes... I also had this consideration initially...

> In theory, here we might not have sent a CloseFrame just yet if the remote initiated the close, but this will call write_pending which will answer the close frame, 

Yeah, but... this won't trigger an error in this case. If remote initiated close and we have not yet received a frame from remote, it's valid to send frames from our side (we don't know that we're going to receive a close frame in a millisecond).

> so it's probably best that the API hides that fact and pretends that we respond immediately to keep things consistent.
> That does mean that client code now really has to check for this in sender tasks, because it can happen right after receiving a close frame, but before receiving ConnectionClosed.

Sorry, I did not get it.

Reply:
> Yes, that looks good for me.
I updated the PR with this proposition

> If remote initiated close and we have not yet received a frame from remote...

I meant if we have received it, but our response is still sitting in the send queue. It is sound as long as all user facing API's consider it as if we had already sent it out, since any new user messages will end up behind it in the queue. However all internal code must be careful to not consider the close handshake finished, since it isn't really and we should never set state to terminated and return ConnectionClosed before it's actually sent out. That works correctly now, but people modifying the code in the future should be aware of this, and it's not really documented anywhere.

As far as the Protocol error now appearing in `write_message`, it might break client code because they might be sending messages in a loop and matching for `SendQueueFull` and `ConnectionClosed`, however now they might get `Protocol`. So it's a breaking change and I would mention this in release notes.

Reply:
The test is failing. I'm working on it, don't merge.

Reply:
Ok, tests pass. 

There is a scenario to be aware of:
1. Remote initiate close handshake
2. we receive it and cue a response
3. we try to send -> protocol error
4. client code here must still do calls to `write_pending` until the close response is out. If the connection get's dropped right after the protocol error, the other side will receive a `Protocol( connection reset )` error.

Reply:
> There is a scenario to be aware of:

The step (3) in such client code seems to be a bug on the client side, right? (the client tries to send a message after receiving a close message).

Reply:
The thing is you might have sender and reader tasks running concurrently, so unless you have a channel between both in order to immediately intercept the writer, the error messages returned on the writer are meaningful for detecting connection closes and such.

I haven't looked at the code using tungstenite in the wild but I would assume people use the errors for that information to avoid the extra synchronization.